sdk/src/activity_context.rs

```diff
@@ -0,0 +1,202 @@
+    pub(crate) input: Vec<Payload>,
+    pub(crate) heartbeat_details: Vec<Payload>,
+    /// #NOTE future usage?
+    pub(crate) header_fields: HashMap<String, Payload>,
+    pub(crate) info: ActivityInfo,
+}
+
+#[derive(Clone)]
+pub struct ActivityInfo {
+    pub task_token: Vec<u8>,
+    pub workflow_type: String,
+    pub workflow_namespace: String,
+    pub workflow_execution: Option<WorkflowExecution>,
+    pub activity_id: String,
+    pub activity_type: String,
+    pub task_queue: String,
+    pub heartbeat_timeout: Option<Duration>,
+    /// Time of activity scheduled by a workflow
+    pub scheduled_time: Option<Timestamp>,
+    /// Time of activity start
+    pub started_time: Option<Timestamp>,
+    /// Time of activity timeout
+    pub deadline: Option<Timestamp>,
+    /// Attempt starts from 1, and increased by 1 for every retry if retry policy is specified.
+    pub attempt: u32,
+
+    /// #NOTE Not in sdk-go
+    pub current_attempt_scheduled_time: Option<Timestamp>,
+    pub retry_policy: Option<RetryPolicy>,
+    pub is_local: bool,
+}
+
+impl ActContext {
+    /// Construct new Activity Context
+    pub fn new(
+        worker: Arc<dyn Worker>,
+        task_queue: String,
+        task_token: Vec<u8>,
+        task: activity_task::Start,
+    ) -> Self {
+        let activity_task::Start {
+            workflow_namespace,
+            workflow_type,
+            workflow_execution,
+            activity_id,
+            activity_type,
+            header_fields,
+            input,
+            heartbeat_details,
+            scheduled_time,
+            current_attempt_scheduled_time,
+            started_time,
+            attempt,
+            schedule_to_close_timeout,
+            start_to_close_timeout,
+            heartbeat_timeout,
+            retry_policy,
+            is_local,
+        } = task;
+        let deadline = calculate_deadline(
+            scheduled_time.as_ref(),
+            started_time.as_ref(),
+            start_to_close_timeout.as_ref(),
+            schedule_to_close_timeout.as_ref(),
+        );
+
+        ActContext {
+            worker,
+            input,
+            heartbeat_details,
+            header_fields,
+            info: ActivityInfo {
+                task_token,
+                task_queue,
+                workflow_type,
+                workflow_namespace,
+                workflow_execution,
+                activity_id,
+                activity_type,
+                heartbeat_timeout,
+                scheduled_time,
+                started_time,
+                deadline,
+                attempt,
+
+                current_attempt_scheduled_time,
+                retry_policy,
+                is_local,
+            },
+        }
+    }
+
+    /// Retrieve the rest of the parameters
+    /// #NOTE the first parameter is popped for convenience, the rest is still stored in the
+    /// activity state.  There are no variadic parameters in Rust.
```

Comment:
This is probably a bit confusing. I think I want to continue enforcing that Rust-defined activities/workflows only take 1 parameter.

In that case, we should maybe give it a more specific name like `extra_inputs` . The docstring can indicate the first parameter is passed to the activity function explicitly, and there should not be any extra unless the caller is doing something a bit odd by passing more than one, and those can be retrieved here.

Reply:
Fixed by renaming + docs.  I think it's healthy to keep the function even if a little confusing because it is possible to have polyglot environments where workflows are deployed from Go, but certain activities are handled in Rust specific task_queues.  In that case, it could be surprising if the Go SDK naturally allows many parameters but Rust's activities can never access them.